### ISSUE TITLE
Is it possible for a new node to cause duplicate activation when it joins?

### ISSUE THREAD

**Issue by liuhan907** [dotnet/orleans#5112]
When I read the Orleans code, 
I found that when a new node joins the cluster and when its state becomes active, 
the other nodes in the cluster call the LocalGrainDirectory's SiloStatusChangeNotification method.
The 461 row of this file will call the AddServer method to handle the new node join.

But when I read the Addserver method, 
I found that the 309 lines in this file would call the GrainDirectoryHandoffManager's ProcessSiloAddEvent method. 
This method will check whether the newly joined node is its immediate successor. 
If so, split self GrainDirectory and submit the part that belongs to the new node to it.

However, the problem here is that when a node state becomes active, 
it can accept messages sent by other nodes.
Then when the split GrainDirectory data is submitted to the new node, 
the other nodes send an activation request to the new node, 
which can cause duplicate activation.

Let me give you an example, 
we have three nodes: B, C, and D.
In node C activated Grain[A-1], the activation information is saved on Node B. 
This time the new node A joins the cluster, node A is the direct forward of B, 
and the information of Grain[A-1] now belongs to A.
At this point, when the status of node A becomes active, 
node B begins to process the AddServer event and prepares to give the Grain[A-1] location data that it retains to node A. 
But at this point, node D generates a call to Grain[A-1], 
at which point node D will see if the location information is on node A, 
but because node B has not given the data to node A, 
so node D found no grain[A-1] location information, 
then according to the default policy, will randomly select a node to activate,
Assuming node D is selected, then D will have a grain[A-1] activation. 
The cluster will now have two grain[A-1] activations.

Is this situation possible or impossible? My understanding of the grain activation process is wrong?

**Comment by benjaminpetit** [comment]
Yes you are correct, it is possible to have two activations for the same grain id at a given time, in the scenario you describe.

But, when the cluster will "stabilize" (in your example when the handoff will occur), one activation will "win": the directory will always returns the same activation address for a given grain id. 

And this should happen very quickly.

The grain that lost the race will not be deactivated right away, but it will be isolated and should not receive requests anymore. I believe that if you have timer registered (timers, not reminders) then the grain that lost will still receive them. But timers do not reset the idle time, so the grain will be naturally deactivated by the garbage collector at some point.

We have some planned work to make the directlroy handoff more robust, that will probably be done for 3.0.

Don't hesitate if you have more questions.

**Comment by liuhan907** [comment]
Thank you.
Another question is, what are the ideas if you want to make the handoff more robust? 
My idea is to make silo not accept application messages before the data handoff is complete, 
but I worry that in the middle of the handoff, 
if the handoff node crashes, no one will be able to tell the new node that the handoff is complete.

**Comment by sergeybykov** [comment]
@ReubenBond was previously looking into the possibility of providing a stronger single-activation guarantee and making grain directory more robust in general. Until we invest into that, the possibility of temporarily having multiple activations is the reality.

**Comment by liuhan907** [comment]
Thanks, I don't have any more questions for a while.

**Comment by benjaminpetit** [comment]
But then what happen when two silos join the cluster at the same time? How do you know for sure which silo was the predecessor?

**Comment by sergeybykov** [comment]
@liuhan907 Different nodes in the cluster may have different views of the cluster state during cluster changes. Because of that, it is impossible to ensure that two (or more) of them always agree on which silo should hold a directory partition for a grain without a more elaborate protocol, e.g. with leases.

**Comment by liuhan907** [comment]
Oh, yes, I didn't consider adding a plural node at the same time.
